tools/codegen/dest/ufunc.py

```diff
@@ -0,0 +1,440 @@
+            if isinstance(b.argument, Argument) and b.argument.type != BaseType(BaseTy.Scalar):
+                continue
+            body.append(f"auto _v_{b.name} = at::vec::Vectorized<scalar_t>(_s_{b.name});")
+            ctx.append(Expr(f"_v_{b.name}", NamedCType(b.nctype.name, VectorizedCType(BaseCType(scalar_t)))))
+
+    # Setup lambda signature
+    # NB: simplified version of ufunctor_arguments
+    scalar_bindings = []
+    vec_bindings = []
+    for a in g.functional.func.arguments.flat_non_out:
+        if not a.type.is_tensor_like():
+            continue
+        assert a.type == BaseType(BaseTy.Tensor)
+        scalar_bindings.append(Binding(
+            name=a.name,
+            nctype=NamedCType(a.name, BaseCType(scalar_t)),
+            argument=a,
+        ))
+        if vec_loop is not None:
+            vec_bindings.append(Binding(
+                name=a.name,
+                nctype=NamedCType(a.name, VectorizedCType(BaseCType(scalar_t))),
+                argument=a,
+            ))
+
+    def with_ctx(b: Sequence[Binding]) -> List[Union[Expr, Binding]]:
+        r: List[Union[Expr, Binding]] = []
+        r.extend(ctx)
+        r.extend(b)
+        return r
+
+    body_str = '\n'.join(body)
+    if vec_loop is not None:
+        return f"""
+{body_str}
+cpu_kernel_vec(iter,
+  [=]({', '.join(b.decl() for b in scalar_bindings)}) {{ return {scalar_loop.call(with_ctx(scalar_bindings))}; }},
+  [=]({', '.join(b.decl() for b in vec_bindings)}) {{ return {vec_loop.call(with_ctx(vec_bindings))}; }}
+);
+"""
+    else:
+        return f"""
+{body_str}
+cpu_kernel(iter,
+  [=]({', '.join(b.decl() for b in scalar_bindings)}) {{ return {scalar_loop.call(with_ctx(scalar_bindings))}; }}
+);
+"""
+
+@with_native_function
+def compute_ufunc_cpu_kernel(g: NativeFunctionsGroup) -> str:
+    stub_sig = StubSignature(g)
+
+    # Reindex the ufunc by dtypes; processing generic/scalaronly as well
+    loops = g.out.ufunc_inner_loop
+    ufunc_sigs: Dict[ScalarType, Dict[UfuncKey, UfuncSignature]] = {}
+    for k in [UfuncKey.CPUScalar, UfuncKey.CPUVector]:
+        lks = []
+        # ORDER MATTERS: this specifies overriding precedence
+        if k in loops:  # should happen rarely
+            lks.append(k)
+        if UfuncKey.ScalarOnly in loops and k is UfuncKey.CPUScalar:
+            lks.append(UfuncKey.ScalarOnly)
+        if UfuncKey.Generic in loops:
+            lks.append(UfuncKey.Generic)
+        # TODO: don't hardcode ufunc:: namespace here, should be centralized smh
+        for lk in lks:
+            for dtype in loops[lk].supported_dtypes:
+                compute_t: CType
+                if k is UfuncKey.CPUScalar:
+                    compute_t = BaseCType(scalar_t)
+                elif k is UfuncKey.CPUVector:
+                    compute_t = VectorizedCType(BaseCType(scalar_t))
+                else:
+                    assert False
+                inner_ufunc_sigs = ufunc_sigs.setdefault(dtype, {})
```

Comment:
I might be reading this wrong - say a user defines their yaml as:
```
  ufunc_inner_loop:
    Generic: add (AllAndComplex)
    ScalarOnly: add (Bool)
```

Even though `Bool` is technically part of the `AllAndComplex` set, we want to treat `ScalarOnly` specially to say "generate a scalar kernel but not a vector kernel for `bool`".

I'm probably just reading this wrong - what part of the logic here prevents us from generating a vector kernel for `bool` in that case? It looks like we generate a `Dict[ScalarType, Dict[UfuncKey, UfuncSignature]]`, and for `bool`, we'll just add a separate entry for `CPUScalar` and `CPUVector`.

Reply:
It's actually not; here are the classes:

```
+DTYPE_CLASSES: Dict[str, Set[ScalarType]] = {}
+# NB: Integral doesn't include boolean
+DTYPE_CLASSES["Integral"] = {
+    ScalarType.Byte, ScalarType.Char, ScalarType.Int, ScalarType.Long,
+    ScalarType.Short
+}
+# NB: Floating doesn't include low precision types
+DTYPE_CLASSES["Floating"] = {ScalarType.Float, ScalarType.Double}
+DTYPE_CLASSES["Complex"] = {ScalarType.ComplexFloat, ScalarType.ComplexDouble}
+DTYPE_CLASSES["All"] = DTYPE_CLASSES["Integral"] | DTYPE_CLASSES["Floating"]
+DTYPE_CLASSES["AllAndComplex"] = DTYPE_CLASSES["All"] | DTYPE_CLASSES["Complex"]
+DTYPE_CLASSES["FloatingAndComplex"] = DTYPE_CLASSES["Floating"] | DTYPE_CLASSES["Complex"]
```

Bool is not in any of the precanned classes. So there is no conflict. 

I do agree the logic here is a bit mushy. It's one of my least favorite parts of this PR.

Reply:
So your example doesn't work, but we could look at an example that does work:

```
 ufunc_inner_loop:
    Generic: add (Bool)
    ScalarOnly: add_scalar (Bool)
```

What does this do? Well, you get an error:

```
  File "/private/home/ezyang/pytorch-tmp/tools/codegen/dest/ufunc.py", line 133, in compute_ufunc_cuda_functors
    assert ufunc_name == loops[lk].name, "ScalarOnly and Generic must have same ufunc name"
AssertionError: ScalarOnly and Generic must have same ufunc name
```

If the ufunc names are the same, there's no error, but you will get the generic implementation with scalar+vectorized (because you said it was OK). Probably a good idea to give an error in this case, though a little complex as I will need to implement a subsumption check (e.g., "this line is useless because it is completely overwritten by some other line")

Reply:
For now I'm not going to implement the shadowing logic but it is tracked with an expected failing test case at test_scalaronly_shadowed (to be pushed shortly)